### PR TITLE
quantpass bugfixes

### PR DESCRIPTION
**Pull request by swhite2401**
This PR is a follow-up of #502. It was found that `*QuantPass.c` passmethods had erratic behavior and were producing wrong results.

Few bugs were identified and are fixed in the proposal. 
A more detailed code review and testing might be necessary for validation.

### CODE REVIEW

**Comment by swhite2401** [comment]
@JonasKallestrup I cannot add you as reviewer but it would be nice if you could take a look too

**Comment by swhite2401** [comment]
First benchmarking test:

Comparison between (python syntax):
```
ohmi_envelop(ring.radiation_on())
lattice_pass(ring.radiation_on()+[quantdiff_elem])
lattice_pass(ring.radiation_on(dipole_pass='BndMPoleSymplectic4QuantPass', quadrupole_pass='StrMPoleSymplecticQuantPass')
```
This was done for the standard ESRF-EBS S28F lattice.

image

**Comment by swhite2401** [comment]
Bencmarking continued:
This cases shows the energy loss after one quadrupole, in this case a very (unrealistic) strong quadrupole was used to increase the probability for photon emission

image

**Comment by swhite2401** [comment]
Last verification:
Now I apply a frequency shift equivalent to a shift of dp/p of 1%, the following are expected:
-a reduction of horizontal emittance w.r.t. on-energy
-an increase of long. emittance w.r.t. on-energy
-> these are driven by the dispersive orbit in quadrupoles, this a good check that radiation in quadrupoles is properly implemented

I still track 1000 particles, the slicing of quadrupoles was increased and RFCavityPass has to be used.

image

**Comment by swhite2401** [comment]
**Conclusions**: 
- There were several bugs in the present implementation
- Consistency checks of individual magnets and full ring simulation have been performed after the fixes
- All simulations show good agreement with expectations

**For the moment everything looks correct to me.** Comments and additional tests more than welcome

**Comment by lfarv** [comment]
I tested the GetEnergy function by building the histogram of 10000 calls:


getenergy

The result is perfect: the generated values of photon energy follow the expected value

**Comment by lfarv** [comment]
The threshold of 0.21 modified in the PR is the low limit of the tabulated value for photon energy. I looked at the behaviour around that point: below 0.21, the energy is computed analytically (21 % of the emitted photons), above it is interpolated in a table (79 % of the photons). The high end is used for only 1/100000 of the photons !



discontinuity

There is a small gap in energy at &epsilon;/&epsilon;<sub>0</sub> = 0.005, with a width of 0.00014. This is negligible and does not appear in the histogram (see previous comment).

So the generation of photon energy is correct !

File: atintegrators/BndMPoleSymplectic4QuantPass.c [hunk 1]
```diff
@@ -123,27 +123,25 @@ void BndMPoleSymplectic4QuantPass(double *r, double le, double irho, double *A,
...
                 de = 0.0;
                 if(nph>0){
                     for(i=0;i<nph;i++){
                         de = de + getEnergy(ec);
                     };
-                };
-                
-                r6[1] = r6[1]/(1+r6[4])*(1+r6[4]-de/E0);
-                r6[3] = r6[3]/(1+r6[4])*(1+r6[4]-de/E0);
+                };
```
Comment (lfarv) [inline atintegrators/BndMPoleSymplectic4QuantPass.c:141]: Just a detail: the first test is redundant: the loop will not execute if the initial condition is not fulfilled…

**Review by lfarv — APPROVED** [review]

File: atintegrators/BndMPoleSymplectic4QuantPass.c (hunk 1 above)
Comment (swhite2401) [reply atintegrators/BndMPoleSymplectic4QuantPass.c]: Right, corrected

**Comment by swhite2401** [comment]
@lfarv, should we use this branch to provide on/off features?

**Comment by lfarv** [comment]
@swhite2401 : I think we can merge this now: it solves a real bug, while quantum diffusion can still be turned on/off using `enable_6d`/`disable_6d`, even if it is not so convenient.

For the following, I even see 2 steps:
1. Correcting the fact that `check_6d` and `disable_6d` ignore the `QuantPass` suffix. It's a potentially dangerous bug,
2. Find the best way to  enable local quantum radiation, either with a dedicated function, of by adapting `enable_6d`. For instance one could imagine replacing the 'auto' code which selects the adequate pass method by either 'rad' for "classical" radiation or 'quantum' for quantum radiation. A dedicated function might be simple to use, I just would try to avoid duplicating the `enable_6d` code…

**Comment by lfarv** [comment]
I have a doubt concerning OpenMP, or other parallel execution: will the `rand()` function generate the same sequence in the different threads (which would be bad), or still generate a new number on each call, possibly from different threads ? It's not clear in the POSIX documentation, and may depend on the implementation. GNU libc specifies that `rand` is thread safe, is it enough ?

**Comment by swhite2401** [comment]
> I have a doubt concerning OpenMP, or other parallel execution: will the `rand()` function generate the same sequence in the different threads (which would be bad), or still generate a new number on each call, possibly from different threads ? It's not clear in the POSIX documentation, and may depend on the implementation. GNU libc specifies that `rand` is thread safe, is it enough ?

Very good point! I am afraid we have to initialize the random number generator with different seeds in each thread....that is quite a bit of a change w.r.t. the present implementation

**Comment by swhite2401** [comment]
@lfarv so I definitely confirm this is a problem, and not only for this part but for functions using random number generation:
-particles distribution
-lumped diffusion element
-element by element photon emission
-others?

and this for all parallelized processes: MPI/OpenMP/python multiprocessing

The fix for each case is different and will have to take place in different parts of the code, but generally speaking we have have to produce a different seed for each thread.

The most difficult case at first sight is the python multiprocessing as I have no idea on how to pass the process number to the C engine in this case...

What I propose is to merge this branch as is because this is self contained and fixes obvious bugs and then make a separate PR to handle random number generation for parallelized applications

What do you think?

**Comment by lfarv** [comment]
@swhite2401 : I agree to merge this branch right now.

It would be useful to make some tests to see what happens for different parallel methods and rand() implementations, but if looks not so simple…

**Comment by JonasKallestrup** [comment]
> @JonasKallestrup I cannot add you as reviewer but it would be nice if you could take a look too

I've done the tracking simulation with the SLS 2 lattice (i.e. with anti-bends) and with a strong dispersive coupling source, and it results in the correct transverse emittances.

**Comment by swhite2401** [comment]
Yes I fully agree testing is necessary but complicated...

Thanks both for your feedback, I merge this as is then

**Pull request merged**